### PR TITLE
Fix modal vs modal-button discrepancy in resource gallery

### LINKED ISSUE #235: Resource Gallery "See More" opens incorrect modal
On the resource gallery, when I click the "... More" to open the modal with the full card description, it opens the incorrect modal. I so far only noticed this with the "An Introduction to Earth and Environmental Data Science" card, which opens the Unidata Python Training modal, instead of the expected modal

image

image

I haven't yet gone through the gallery to see how pervasive this is, but I'll comment an update.

### PR DESCRIPTION
**Pull request by jukent**
There are more modals than modal-buttons, so the modals that line up do not match the card from which they are opened.

Closes #235

### CODE REVIEW

File: portal/_extensions/truncatehtml.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def next_token(self):
         try:
             char = self.input[self.counter]
             self.counter += 1
-            if char == '&':
+            if char == '&amp;':
```
Comment (kmpaul) [inline portal/_extensions/truncatehtml.py:82]: This won't work. The `char` is always a single character so this check will always fail.  I would not change this back:

```suggestion
            if char == '&':
```

**Review by kmpaul — CHANGES_REQUESTED** [review]

**Comment by kmpaul** [comment]
This fix has already made some improvements.  In the preview link, I can see that there are 13 `class="modal"` objects and 12 `class="modal-btn"` objects.  So, it's getting better.

**Comment by kmpaul** [comment]
> Interestingly enough, the above fix seems to have fixed the truncated Unidata card description. Not sure why. Maybe the original problematic rendering was a fluke and not reproducible.

I've been thinking about this fact a bit, and I'm bothered by it.  It seems that the problem with the description in the Unidata card that created the original modal mismatches (i.e., not having the `<a class="modal-btn">` tag in it) cannot be reproduced.  I still don't know why this happened, and I'm particularly afraid that it was a fluke and cannot be reproduced!

The reason I'm worried about this is because if an unreproducible error adding the modal button ever happens again (for whatever reason), then the gallery modals will get out of sync again.  I think there are 2 solutions to this problem:

1. Change the `custom.js` script functions to use a named *key*-based indexing (as you pointed out in (https://github.com/ProjectPythia/projectpythia.github.io/issues/235#issuecomment-1118858684).  But I'm not sure this is the best solution because it leaves `class="modal"` objects in the HTML without matching `class="modal-btn"` objects that bloats the HTML (even if the gallery modals still work properly).

2. Change the gallery generator extension to check to see if the `short_description` has the `ellipsis` string in it, such as changing:

   ```python
   if _len_html(item['description']) < max_descr_len:
       short_description = item['description']
       modal_str = ''
   else:
       short_description = th.truncate(
           item['description'], max_descr_len, ellipsis='<a class="modal-btn"> ...more</a>'
       )
       modal_str = f"""
   # ...contents removed for brevity...
   """
   ```

   to:

   ```python
   ellipsis_str = '<a class="modal-btn"> ...more</a>'
   short_description = th.truncate(item['description'], max_descr_len, ellipsis=ellipsis_str)
   if ellipsis_str in short_description:
       modal_str = f"""
   # ...contents cut for brevity...
   """
   else:
       modal_str = ''
   ```

   This solution *always* truncates the description string using the `truncate` function.  And it doesn't need the `_len_html` function at all.  However, if the description string is shorter than the `max_descr_len` (according to the `truncate` function), then it will add the `ellipsis_str` string to the end.  And *only if the `ellipsis_str` is added* does it create the `class="modal"` object.  

   This means that even if the `truncate` function fails for some mysterious reason, and it fails to truncate or something else weird happens, then it only creates the modal if the modal button actually exists.  I think this is safer.

**Comment by kmpaul** [comment]
> Interestingly enough, the above fix seems to have fixed the truncated Unidata card description. Not sure why. Maybe the original problematic rendering was a fluke and not reproducible.

I've been thinking about this fact a bit, and I'm bothered by it.  It seems that the problem with the description in the Unidata card that created the original modal mismatches (i.e., not having the `<a class="modal-btn">` tag in it) cannot be reproduced.  I still don't know why this happened, and I'm particularly afraid that it was a fluke and cannot be reproduced!

The reason I'm worried about this is because if an unreproducible error adding the modal button ever happens again (for whatever reason), then the gallery modals will get out of sync again.  I think there are 2 solutions to this problem:

1. Change the `custom.js` script functions to use a named *key*-based indexing (as you pointed out in (https://github.com/ProjectPythia/projectpythia.github.io/issues/235#issuecomment-1118858684).  But I'm not sure this is the best solution because it leaves `class="modal"` objects in the HTML without matching `class="modal-btn"` objects that bloats the HTML (even if the gallery modals still work properly).

2. Change the gallery generator extension to check to see if the `short_description` has the `ellipsis` string in it, such as changing:

   ```python
   if _len_html(item['description']) < max_descr_len:
       short_description = item['description']
       modal_str = ''
   else:
       short_description = th.truncate(
           item['description'], max_descr_len, ellipsis='<a class="modal-btn"> ...more</a>'
       )
       modal_str = f"""
   # ...contents removed for brevity...
   """
   ```

   to:

   ```python
   ellipsis_str = '<a class="modal-btn"> ...more</a>'
   short_description = th.truncate(item['description'], max_descr_len, ellipsis=ellipsis_str)
   if ellipsis_str in short_description:
       modal_str = f"""
   # ...contents cut for brevity...
   """
   else:
       modal_str = ''
   ```

   This solution *always* truncates the description string using the `truncate` function.  And it doesn't need the `_len_html` function at all.  However, if the description string is shorter than the `max_descr_len` (according to the `truncate` function), then it will add the `ellipsis_str` string to the end.  And *only if the `ellipsis_str` is added* does it create the `class="modal"` object.  

   This means that even if the `truncate` function fails for some mysterious reason, and it fails to truncate or something else weird happens, then it only creates the modal if the modal button actually exists.  I think this is safer.

**Comment by jukent** [comment]
Will address the shortening of the Unidata card in a separate PR.

**Review by kmpaul — APPROVED** [review]
Thanks for doing this, @jukent!

**Review by mgrover1 — APPROVED** [review]
Fantastic work @jukent !!

**Pull request merged**